src/libmw/src/wallet/Transact.cpp

```diff
@@ -0,0 +1,123 @@
+    );
+
+    // Get input coins
+    LOG_INFO_F(
+        "Creating Txs: Inputs({}), pegins({}), pegouts({}), recipients({}), fee({})",
+        WalletUtil::TotalAmount(input_coins),
+        pegin_amount.value_or(0),
+        pegout_total,
+        recipient_total,
+        fee
+    );
+    if ((WalletUtil::TotalAmount(input_coins) + pegin_amount.value_or(0)) != (pegout_total + recipient_total + fee)) {
+        ThrowInsufficientFunds("Total amount mismatch");
+    }
+
+    // Sign inputs
+    std::vector<Input> inputs = WalletUtil::SignInputs(input_coins);
+
+    // Create outputs
+    Transact::Outputs outputs = CreateOutputs(recipients);
+
+    // Total kernel offset is split between raw kernel_offset and the kernel's blinding factor.
+    // sum(output.blind) - sum(input.blind) = kernel_offset + sum(kernel.blind)
+    std::vector<BlindingFactor> input_blinds = WalletUtil::GetBlindingFactors(input_coins);
+    BlindingFactor kernel_offset = Random::CSPRNG<32>();
+    BlindingFactor kernel_blind = Blinds()
+        .Add(outputs.total_blind)
+        .Sub(input_blinds)
+        .Sub(kernel_offset)
+        .Total();
+
+    // Create the kernel
+    Kernel kernel = Kernel::Create(
+        kernel_blind,
+        fee,
+        pegin_amount,
+        pegouts.empty() ? boost::none : boost::make_optional(pegouts.front()),
+        boost::none
+    );
+
+    // FUTURE: Only necessary when none of the addresses are owned by this wallet?
+    BlindingFactor owner_sig_key = Random::CSPRNG<32>();
+    SignedMessage owner_sig = Schnorr::SignMessage(owner_sig_key.GetBigInt(), kernel.GetHash());
+
+    // Total owner offset is split between raw owner_offset and the owner_sig's key.
+    // sum(output.sender_key) - sum(input.key) = owner_offset + sum(owner_sig.key)
+    std::vector<BlindingFactor> input_keys = WalletUtil::GetKeys(input_coins);
+    BlindingFactor owner_offset = Blinds()
+        .Add(outputs.total_key)
+        .Sub(input_keys)
+        .Sub(owner_sig_key)
+        .Total();
+
+    // Build the transaction
+    return mw::Transaction::Create(
+        kernel_offset,
+        owner_offset,
+        inputs,
+        std::move(outputs.outputs),
+        std::vector<Kernel>{ std::move(kernel) },
+        std::vector<SignedMessage>{ std::move(owner_sig) }
+    );
+}
+
+Transact::Outputs Transact::CreateOutputs(const std::vector<std::pair<uint64_t, StealthAddress>>& recipients)
+{
+    Blinds output_blinds;
+    Blinds output_keys;
+    std::vector<Output> outputs;
+    std::transform(
+        recipients.cbegin(), recipients.cend(),
+        std::back_inserter(outputs),
+        [&output_blinds, &output_keys](const std::pair<uint64_t, StealthAddress>& recipient) {
+            BlindingFactor blind;
+            SecretKey ephemeral_key = Random::CSPRNG<32>();
+            Output output = Output::Create(
+                blind,
+                EOutputFeatures::DEFAULT_OUTPUT,
```

Comment:
Even for pegged-in outputs?

I noticed that EOutputFeatures::PEGGED_IN isn't set anywhere in the code.